src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs

```diff
@@ -315,84 +314,143 @@ private static bool AreCompatible(DynamicallyAccessedMemberTypes serviceDynamica
-                            // so we need to ask creation with the original identity of the descriptor
-                            var registrationKey = isAnyKeyLookup ? ServiceIdentifier.FromDescriptor(_descriptors[i]) : cacheKey;
-                            if (TryCreateExact(_descriptors[i], registrationKey, callSiteChain, slot) is { } callSite)
+                            // so we need to ask creation with the original identity of the descriptor.
+                            registrationKey = isAnyKeyLookup ? ServiceIdentifier.FromDescriptor(_descriptors[i]) : cacheKey;
+
+                            slot = GetSlot();
+                            if (enumerationMode == EnumerationMode.Exact)
                             {
+                                ServiceCallSite callSite = CreateExact(_descriptors[i], registrationKey, callSiteChain, slot);
+                                Debug.Assert(callSite != null);
                                 AddCallSite(callSite, i);
                             }
-                        }
-                    }
-                    for (int i = _descriptors.Length - 1; i >= 0; i--)
-                    {
-                        if (KeysMatch(cacheKey.ServiceKey, _descriptors[i].ServiceKey))
-                        {
-                            // Special case for AnyKey: we don't want to add in cache a mapping AnyKey -> specific type,
-                            // so we need to ask creation with the original identity of the descriptor
-                            var registrationKey = isAnyKeyLookup ? ServiceIdentifier.FromDescriptor(_descriptors[i]) : cacheKey;
-                            if (TryCreateOpenGeneric(_descriptors[i], registrationKey, callSiteChain, slot, throwOnConstraintViolation: false) is { } callSite)
+                            else if (CreateOpenGeneric(_descriptors[i], registrationKey, callSiteChain, slot, throwOnConstraintViolation: false) is { } callSite)
                             {
                                 AddCallSite(callSite, i);
                             }
                         }
-                    }
 
-                    callSitesByIndex.Sort((a, b) => a.Key.CompareTo(b.Key));
-                    callSites = new ServiceCallSite[callSitesByIndex.Count];
-                    for (var i = 0; i < callSites.Length; ++i)
-                    {
-                        callSites[i] = callSitesByIndex[i].Value;
-                    }
+                        void AddCallSite(ServiceCallSite callSite, int index)
+                        {
+                            cacheLocation = GetCommonCacheLocation(cacheLocation, callSite.Cache.Location);
+                            callSitesByIndex.Add(new(index, callSite));
+                            UpdateSlot();
+                        }
 
-                    void AddCallSite(ServiceCallSite callSite, int index)
-                    {
-                        slot++;
+                        int GetSlot()
```

Comment:
>int GetSlot() [](http://example.com/codeflow?start=24&length=13)

can we make this statis and pass the needed parameters?

Reply:
I think keeping it as-is helps since there are 5 variables (isAnyKeyLookup, slot, keyedSlotAssignment, _descriptors, registrationKey) and tight coupling here.